gflownet/envs/crystals/miller.py

```diff
@@ -28,13 +33,58 @@ class MillerIndices(Grid):
 
     def __init__(
         self,
-        is_cubic: bool,
+        is_hexagonal_rhombohedral: bool,
         max_increment: int = 1,
         max_dim_per_action: int = 1,
         **kwargs,
     ):
-        if is_cubic:
-            n_dim = 3
-        else:
-            n_dim = 4
-        super().__init__(n_dim=n_dim, length=5, cell_min=-2, cell_max=2, **kwargs)
+        self.is_hexagonal_rhombohedral = is_hexagonal_rhombohedral
+        super().__init__(n_dim=3, length=5, cell_min=-2, cell_max=2, **kwargs)
+
+    def get_mask_invalid_actions_forward(
+        self,
+        state: Optional[List] = None,
+        done: Optional[bool] = None,
+    ) -> List:
+        """
+        Returns a list of length the action space with values:
+            - True if the forward action is invalid from the current state.
+            - False otherwise.
+        """
+        if state is None:
+            state = self.state.copy()
+        if done is None:
+            done = self.done
+
+        mask = super().get_mask_invalid_actions_forward(state, done)
+
+        # If the structure is hexagonal or rhombohedral, we need -2 <= h + k <= 2
+        # for the state to be valid. This means that we have to enforce constraints
+        # on the mask to ensure that the trajectory doesn't end in an invalid state.
+        if self.is_hexagonal_rhombohedral:
+            # Extract the miller indices from the state
+            h, k, l = self.state2oracle(state)
```

Comment:
`state2oracle` will not exist any longer after https://github.com/alexhernandezgarcia/gflownet/pull/247. It is replaced by `state2proxy`. Still, I think it would be safer to not use these methods here in case they change for some reason.

Reply:
Done.
And I agree with both changes you made.